plugins/cryptolib/cryptolib.cpp

```diff
@@ -528,6 +524,90 @@ CRYPTOLIB_API bool CRYPTOLIB_CALL clPKIVerifySignatureBuff(const char * pkalgori
+//  Simple cache for loaded keys
+//
+//-----------------------------------------------------------------
+class CKeyCache : public CInterface
+{
+private:
+    typedef std::unordered_map<string, Owned<CLoadedKey>> KeyCache;
+    KeyCache keyCache;
+
+public:
+    IMPLEMENT_IINTERFACE;
+
+    CLoadedKey * getInstance(bool isPublic, const char * keyFS, const char * keyBuff, const char * passphrase)
+    {
+        if (isEmptyString(keyFS) && isEmptyString(keyBuff))
+            throw MakeStringException(-1, "Must specify a key filename or provide a key buffer");
+        VStringBuffer searchKey("%s_%s_%s", isEmptyString(keyFS) ? "" : keyFS, isEmptyString(keyBuff) ? "" : keyBuff, isEmptyString(passphrase) ? "" : passphrase);
+        KeyCache::iterator it = keyCache.find(searchKey.str());
+        CLoadedKey * ret = nullptr;
+        if (it != keyCache.end())//exists in cache?
+        {
+            ret = (*it).second;
+        }
+        else
+        {
+            if (!isEmptyString(keyFS))
+            {
+                //Create CLoadedKey from filespec
+                if (isPublic)
+                    ret = loadPublicKeyFromFile(keyFS, passphrase);
+                else
+                    ret = loadPrivateKeyFromFile(keyFS, passphrase);
+            }
+            else
+            {
+                //Create CLoadedKey from key contents
+                if (isPublic)
+                    ret = loadPublicKeyFromMemory(keyBuff, passphrase);
+                else
+                    ret = loadPrivateKeyFromMemory(keyBuff, passphrase);
+            }
+
+            keyCache.emplace(searchKey.str(), ret);
+        }
+        return LINK(ret);
+    }
+};
+
+//----------------------------------------------------------------------------
+// TLS storage of Key cache
+//----------------------------------------------------------------------------
+
+static pthread_key_t  tlsBuffKey;
+static pthread_once_t tlsBuffKeyOnce = PTHREAD_ONCE_INIT;
+
+static void tlsBuffKeyAlloc()
+{
+    pthread_key_create(&tlsBuffKey, nullptr);
+}
+
+static CLoadedKey * queryCachedKey(bool isPublic, const char * keyFS, const char * keyBuff, const char * passphrase)
+{
+    Owned<CKeyCache> * pKC = (Owned<CKeyCache>*)pthread_getspecific(tlsBuffKey);
+    if (!pKC)
+    {
+        int rc = pthread_once(&tlsBuffKeyOnce, tlsBuffKeyAlloc);
+        if (rc)
+            throw MakeStringException(-1, "pthread_once failed %d", rc);
+
+        //create the cache
+        static Owned<CKeyCache> TLSKeyCache;
+        TLSKeyCache.setown(new CKeyCache());
+        rc = pthread_setspecific(tlsBuffKey, (const void *)&TLSKeyCache);
+        if (rc)
+            throw MakeStringException(-1, "pthread_setspecific failed %d", rc);
+        pKC = (Owned<CKeyCache>*)pthread_getspecific(tlsBuffKey);
+    }
+    if (!pKC)
+        throw MakeStringException(-1, "Unable to access TLS cache entry");
+
+    return pKC->get()->getInstance(isPublic, keyFS, keyBuff, passphrase);
```

Comment:
@RussWhitehead  Not thinking looking as to why it's crashing, but you should be using a thread_local variables() to implement this (if you search the source, there are various uses).
The impl. should be a lot simpler than above.

NB: you should use addThreadTermFunc to handle clearup.

The crash might be related to the code above, so I suspect worth revisiting that (if it still happens) once this is using thread_local.